packages/filebrowser-extension/src/index.ts

```diff
@@ -533,74 +539,180 @@ function addCommands(
     label: 'New Launcher',
     execute: () => createLauncher(commands, browser)
   });
-}
 
-/**
- * Create a context menu for the file browser listing.
- *
- * #### Notes
- * This function generates temporary commands with an incremented name. These
- * commands are disposed when the menu itself is disposed.
- */
-function createContextMenu(
-  model: Contents.IModel | undefined,
-  commands: CommandRegistry,
-  registry: DocumentRegistry
-): Menu {
-  const menu = new Menu({ commands });
-
-  // If the user did not click on any file, we still want to show
-  // paste as a possibility.
-  if (!model) {
-    menu.addItem({ command: CommandIDs.paste });
-    return menu;
-  }
+  /**
+   * A menu widget that dynamically populates with different widget factories
+   * based on current filebrowser selection.
+   */
+  class OpenwithMenu extends Menu {
+    protected onBeforeAttach(msg: Message): void {
+      // clear the current menu items
+      this.clearItems();
+
+      // get the widget factories that could be used to open all of the items in the current filebrowser selection
+      let factories = OpenwithMenu._intersection(
+        map(tracker.currentWidget.selectedItems(), i => {
+          return OpenwithMenu._getFactories(i);
+        })
+      );
 
-  menu.addItem({ command: CommandIDs.open });
-
-  const path = model.path;
-  if (model.type !== 'directory') {
-    const factories = registry.preferredWidgetFactories(path).map(f => f.name);
-    const notebookFactory = registry.getWidgetFactory('notebook').name;
-    if (
-      model.type === 'notebook' &&
-      factories.indexOf(notebookFactory) === -1
-    ) {
-      factories.unshift(notebookFactory);
+      if (factories) {
+        // make new menu items from the widget factories
+        factories.forEach(factory => {
+          this.addItem({
+            args: { factory: factory },
+            command: CommandIDs.open
+          });
+        });
+      }
+
+      super.onBeforeAttach(msg);
+    }
+
+    static _getFactories(item: Contents.IModel): Array<string> {
+      let factories = registry
+        .preferredWidgetFactories(item.path)
+        .map(f => f.name);
+      const notebookFactory = registry.getWidgetFactory('notebook').name;
+      if (
+        item.type === 'notebook' &&
+        factories.indexOf(notebookFactory) === -1
+      ) {
+        factories.unshift(notebookFactory);
+      }
+
+      return factories;
     }
-    if (path && factories.length > 1) {
-      const command = 'docmanager:open';
-      const openWith = new Menu({ commands });
-      openWith.title.label = 'Open With';
-      factories.forEach(factory => {
-        openWith.addItem({ args: { factory, path }, command });
+
+    static _intersection<T>(
+      object: IterableOrArrayLike<IterableOrArrayLike<T>>
+    ): Set<T> | void {
+      // coerce object to iterator and pop the first element
```

Comment:
`_intersection` is definitely more complex than it needs to be (at some point it became the start of an `intersection.ts` PR for `@phosphor/algorithm`). The catch is that it can only be partially simplified, since the signature will actually need to be `_intersection(object: IIterator<Array<string>>)` (ultimately this is down to the `IIterator` returned by `FileBrowser.selectedItems()`). But at least that way it won't be a template method.

Reply:
Ah, you are totally right 😄 . In that case, carry on with the intersection.

Reply:
done